PiT 24.4: flow does not update entry `"type": "module"` in `package.json`

### Description of the bug

Discovered when [testing](https://github.com/vaadin/hilla/issues/2193) `skeleton-starter-hilla-lit-gradle` in PiT, we have realised that flow since 24.1 is adding the `"type": "module"` when creating a new `package.json` but it is not updating existing ones, causing errors like 

```
Startup of Vite failed. Output was:
✘ [ERROR] "@vaadin/hilla-file-router/vite-plugin.js" resolved to an ESM file. ESM file cannot be loaded by `require`. See https://vitejs.dev/guide/troubleshooting.html#this-package-is-esm-only for more details. [plugin externalize-deps]

    node_modules/esbuild/lib/main.js:1373:27:
      1373 │         let result = await callback({
           ╵                            ^
```

[Slack discussion](https://vaadin.slack.com/archives/C065KMY4D2M/p1709901333027389)

### Expected behavior

projects upgraded to 24.4.0.alpha9+ that use the `hilla-file-router` do not fail if they don't have the entry already in their package.json

### Minimal reproducible example

checkout branch https://github.com/vaadin/skeleton-starter-hilla-lit-gradle/tree/update-to-v24.4 and run ./gradlew bootRun twice 

### Versions

- Vaadin / Flow version: 24.4.0.alpha0

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskGeneratePackageJson.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskGeneratePackageJson.java
@@ -46,6 +46,15 @@ public class TaskGeneratePackageJson extends NodeUpdater {
             JsonObject mainContent = getPackageJson();
             modified = updateDefaultDependencies(mainContent);
             if (modified) {
+                if (!mainContent.hasKey("type")
+                        || !mainContent.getString("type").equals("module")) {
+                    mainContent.put("type", "module");
+                    log().info(
+                            """
+                                    Adding package.json type as module to enable ES6 modules which is now required.
+                                    With this change sources need to use 'import' instead of 'require' for imports.
+                                    """);
+                }
                 writePackageFile(mainContent);
             }
         } catch (IOException e) {

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskUpdatePackages.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskUpdatePackages.java
@@ -93,6 +93,15 @@ public class TaskUpdatePackages extends NodeUpdater {
             boolean npmVersionLockingUpdated = lockVersionForNpm(packageJson);
 
             if (modified || npmVersionLockingUpdated) {
+                if (!packageJson.hasKey("type")
+                        || !packageJson.getString("type").equals("module")) {
+                    packageJson.put("type", "module");
+                    log().info(
+                            """
+                                    Adding package.json type as module to enable ES6 modules which is now required.
+                                    With this change sources need to use 'import' instead of 'require' for imports.
+                                    """);
+                }
                 writePackageFile(packageJson);
             }
 

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskUpdatePackagesNpmTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskUpdatePackagesNpmTest.java
@@ -371,6 +371,51 @@ public class TaskUpdatePackagesNpmTest {
     }
 
     @Test
+    public void missingTypeInPackageJson_typeModuleIsAdded()
+            throws IOException {
+        JsonObject packageJson = getOrCreatePackageJson();
+        Assert.assertFalse("No type should be available",
+                packageJson.hasKey("type"));
+
+        createBasicVaadinVersionsJson();
+
+        final TaskUpdatePackages task = createTask(
+                createApplicationDependencies());
+        task.execute();
+
+        // get latest package.json file
+        packageJson = getOrCreatePackageJson();
+
+        Assert.assertTrue("Type should have been addded.",
+                packageJson.hasKey("type"));
+        Assert.assertEquals("Type should be module", "module",
+                packageJson.getString("type"));
+    }
+
+    @Test
+    public void faultyTypeInPackageJson_typeModuleIsAdded() throws IOException {
+        JsonObject packageJson = getOrCreatePackageJson();
+        packageJson.put("type", "commonjs");
+
+        FileUtils.writeStringToFile(new File(npmFolder, PACKAGE_JSON),
+                packageJson.toJson(), StandardCharsets.UTF_8);
+
+        createBasicVaadinVersionsJson();
+
+        final TaskUpdatePackages task = createTask(
+                createApplicationDependencies());
+        task.execute();
+
+        // get latest package.json file
+        packageJson = getOrCreatePackageJson();
+
+        Assert.assertTrue("Type should not have been removed",
+                packageJson.hasKey("type"));
+        Assert.assertEquals("Type should have been updated to 'module'",
+                "module", packageJson.getString("type"));
+    }
+
+    @Test
     public void npmIsInUse_packageJsonVersionIsUpdated_vaadinSectionIsNotChanged()
             throws IOException {
         final JsonObject packageJson = getOrCreatePackageJson();
@@ -456,6 +501,7 @@ public class TaskUpdatePackagesNpmTest {
         packageJson.put("name", "a");
         packageJson.put("license", "b");
         packageJson.put("version", "c");
+        packageJson.put("type", "module");
 
         LinkedHashSet<String> mainKeys = new LinkedHashSet<>(
                 Arrays.asList(packageJson.keys()));

